core/block_validator.go

```diff
@@ -98,7 +124,45 @@ func (v *BlockValidator) ValidateState(block, parent *types.Block, statedb *stat
 	if root := statedb.IntermediateRoot(v.config.IsEIP158(header.Number)); header.Root != root {
 		return fmt.Errorf("invalid merkle root (remote: %x local: %x)", header.Root, root)
 	}
-	return nil
+
+	// Validate 3 roots from block committed on root chain against the received roots
+	// and throw an error if they don't match
+	validate := func() error {
+		tblock, err := v.rootchain.GetBlock(baseCallOpt, block.Difficulty(), block.Number())
+		if err != nil {
+			log.Error("failed to get block from rootchain", "err", err)
+		}
+
+		if header.TxHash != tblock.TransactionsRoot {
+			return fmt.Errorf("transaction root hash mismatch with rootchain block: local %x, rootchain %x", header.TxHash, tblock.TransactionsRoot)
+		}
+
+		if header.Root != tblock.StatesRoot {
+			return fmt.Errorf("state root hash mismatch with rootchain block: local %x, rootchain %x", header.Root, tblock.StatesRoot)
+		}
+
+		if header.ReceiptHash != tblock.ReceiptsRoot {
+			return fmt.Errorf("receipt root hash mismatch with rootchain block: local %x, rootchain %x", header.ReceiptHash, tblock.ReceiptsRoot)
+		}
+		return nil
+	}
+
+	timer := time.NewTimer(0)
+
+	for {
+		select {
+		case <-timer.C:
+			if v.bc.LastSubmittedNumber.Cmp(block.Number()) < 0 && v.bc.LastSubmittedFork.Cmp(block.Difficulty()) < 0 {
```

Comment:
`LastSubmittedNumber` and `LastSubmittedFork` may be nil if block is not submitted. Read it from RootChain contract and save it when `pls` is created.

Reply:
I think `LastSubmittedNumber ` and `LastSubmittedFork` are not necessary since we can check whether a specific block is submitted or not by checking the return value of `rootchain.Getblock()` is empty.

Reply:
But already all `validate` require at least one JSONRPC communication, if we don't do like what I said above. Because `validate` needs to check a validity by comparing a result value of `rootchain.GetBlock()` anyway.